File: lab4and5hw/A2/weather_server.cc
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <random>
#include <algorithm>
#include <stdlib.h>

#include <grpcpp/ext/proto_server_reflection_plugin.h>
#include <grpcpp/grpcpp.h>
#include <grpcpp/health_check_service_interface.h>

#include "weather.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::ServerWriter;
using grpc::Status;
using grpc::StatusCode;
using weather::WeatherSubscription;
using weather::SubscriptionRequest;
using weather::SubscriptionResponse;
using weather::WeatherInformation;

class WeatherImpl final : public WeatherSubscription::Service
{
public:
    WeatherImpl()
    {
        auto weatherInPoland = SubscriptionResponse();
        weatherInPoland.set_country("Poland");
        weatherInPoland.set_temperature(20);
        weatherInPoland.set_humidity(60);

        auto weatherInGermany = SubscriptionResponse();
        weatherInGermany.set_country("Germany");
        weatherInGermany.set_temperature(10);
        weatherInGermany.set_humidity(90);

        auto weatherInSpain = SubscriptionResponse();
        weatherInSpain.set_country("Spain");
        weatherInSpain.set_temperature(35);
        weatherInSpain.set_humidity(20);

        subscriptionResponses_ = {
            weatherInPoland,
            weatherInGermany,
            weatherInSpain
        };
    }

    Status subscribe(ServerContext* context, const SubscriptionRequest* request,
                     ServerWriter<SubscriptionResponse>* response) override
    {
        std::random_device dev;
        std::mt19937 rng(dev());
        std::uniform_int_distribution<std::mt19937::result_type> dist10(1, 5);
        for (const auto& responseWeatherInformation : subscriptionResponses_)
        {
            const auto requestWeatherInformationTypes = request->weather_information().size();
            if (requestWeatherInformationTypes == 0)
            {
                return Status{StatusCode::INVALID_ARGUMENT,
                    "You must set at least one of the weather information types."
                };
            }
            auto satisfiedRequestWeatherInformationTypes = decltype(requestWeatherInformationTypes){0};
            
            for (const auto& requestWeatherInformationType : request->weather_information())
            {
                if (requestWeatherInformationType == WeatherInformation::Temperature)
                {
                    if (request->has_min_temperature() && request->has_max_temperature())
                    {
                        if (request->min_temperature() <= responseWeatherInformation.temperature() &&
                            request->max_temperature() >= responseWeatherInformation.temperature())
                        {
                            ++satisfiedRequestWeatherInformationTypes;
                        }
                        else
                        {
                            break;
                        }
                    }
                    else if (request->has_min_temperature())
                    {
                        if (request->min_temperature() <= responseWeatherInformation.temperature())
                        {
                            ++satisfiedRequestWeatherInformationTypes;
                        }
                        else
                        {
                            break;
                        }
                    }
                    else if (request->has_max_temperature())
                    {
                        if (request->max_temperature() >= responseWeatherInformation.temperature())
                        {
                            ++satisfiedRequestWeatherInformationTypes;
                        }
                        else
                        {
                            break;
                        }
                    }
                    else
                    {
                        return Status{StatusCode::INVALID_ARGUMENT,
                            "You must set at least one of the following: 'min_temperature', 'max_temperature'."
                        };
                    }
                }

                else if (requestWeatherInformationType == WeatherInformation::Humidity)
                {
                    if (request->has_min_humidity() && request->has_max_humidity())
                    {
                        if (request->min_humidity() <= responseWeatherInformation.humidity() &&
                            request->max_humidity() >= responseWeatherInformation.humidity())
                        {
                            ++satisfiedRequestWeatherInformationTypes;
                        }
                        else
                        {
                            break;
                        }
                    }
                    else if (request->has_min_humidity())
                    {
                        if (request->min_humidity() <= responseWeatherInformation.humidity())
                        {
                            ++satisfiedRequestWeatherInformationTypes;
                        }
                        else
                        {
                            break;
                        }
                    }
                    else if (request->has_max_humidity())
                    {
                        if (request->max_humidity() >= responseWeatherInformation.humidity())
                        {
                            ++satisfiedRequestWeatherInformationTypes;
                        }
                        else
                        {
                            break;
                        }
                    }
                    else
                    {
                        return Status{StatusCode::INVALID_ARGUMENT,
                            "You must set at least one of the following: 'min_humidity', 'max_humidity'."
                        };
                    }
                }
            }

            // Simulate response for the server
            sleep(dist10(rng));

            if (requestWeatherInformationTypes == satisfiedRequestWeatherInformationTypes)
            {
                response->Write(responseWeatherInformation);
            }
        }
        return Status::OK;
    }

private:
    std::vector<SubscriptionResponse> subscriptionResponses_;
};

void runServer(const uint16_t port)
{
    std::string serverAddress = "0.0.0.0:" + std::to_string(port);
    WeatherImpl service;

    grpc::EnableDefaultHealthCheckService(true);
    grpc::reflection::InitProtoReflectionServerBuilderPlugin();
    ServerBuilder builder;
    // Listen on the given address without any authentication mechanism.
    builder.AddListeningPort(serverAddress, grpc::InsecureServerCredentials());
    // Register "service" as the instance through which we'll communicate with
    // clients. In this case it corresponds to an *synchronous* service.
    builder.RegisterService(&service);
    // Finally assemble the server.
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "Server listening on " << serverAddress << std::endl;

    // Wait for the server to shutdown. Note that some other thread must be
    // responsible for shutting down the server for this call to ever return.
    server->Wait();
}

int main(int argc, char** argv)
{
    runServer(50051);
    return 0;
}

File: lab4and5hw/I4/gRPC/WeatherServer.java
package io.grpc.examples.weather;

import io.grpc.*;
import io.grpc.stub.StreamObserver;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;
import java.util.concurrent.ThreadLocalRandom;



public class WeatherServer {
    private static final Logger logger = Logger.getLogger(WeatherServer.class.getName());

    private final int port;
    private final Server server;

    /** Create a Weather server using serverBuilder as a base and features as data. */
    public WeatherServer(int port) {
        ServerBuilder<?> serverBuilder = Grpc.newServerBuilderForPort(port, InsecureServerCredentials.create());
        this.port = port;
        server = serverBuilder.addService(new WeatherService(new LinkedList<>()))
                .build();
    }

    /** Start serving requests. */
    public void start() throws IOException {
        server.start();
        logger.info("Server started, listening on " + port);
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            // Use stderr here since the logger may have been reset by its JVM shutdown hook.
            System.err.println("*** shutting down gRPC server since JVM is shutting down");
            try {
                WeatherServer.this.stop();
            } catch (InterruptedException e) {
                e.printStackTrace(System.err);
            }
            System.err.println("*** server shut down");
        }));
    }

    /** Stop serving requests and shutdown resources. */
    public void stop() throws InterruptedException {
        if (server != null) {
            server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
        }
    }

    /**
     * Await termination on the main thread since the grpc library uses daemon threads.
     */
    private void blockUntilShutdown() throws InterruptedException {
        if (server != null) {
            server.awaitTermination();
        }
    }

    /**
     * Main method.  This comment makes the linter happy.
     */
    public static void main(String[] args) throws Exception {
        WeatherServer server = new WeatherServer(50051);
        server.start();
        server.blockUntilShutdown();
    }

    private static class WeatherService extends WeatherSubscriptionGrpc.WeatherSubscriptionImplBase {
        private final Collection<SubscriptionResponse> subscriptionResponses;

        WeatherService(Collection<SubscriptionResponse> subscriptionResponses) {
            this.subscriptionResponses = subscriptionResponses;
            SubscriptionResponse weatherInPoland = SubscriptionResponse.newBuilder().setCountry("Poland").setHumidity(60).build();
            SubscriptionResponse weatherInGermany = SubscriptionResponse.newBuilder().setCountry("Germany").setTemperature(10).build();
            SubscriptionResponse weatherInSpain = SubscriptionResponse.newBuilder().setTemperature(35).setHumidity(20).build();
            subscriptionResponses.add(weatherInPoland);
            subscriptionResponses.add(weatherInGermany);
            subscriptionResponses.add(weatherInSpain);
        }

        @Override
        public void subscribe(SubscriptionRequest request, StreamObserver<SubscriptionResponse> response) {
            for (SubscriptionResponse responseWeatherInformation : subscriptionResponses) {
                int requestWeatherInformationTypes = request.getWeatherInformationCount();
                if (requestWeatherInformationTypes == 0) {
                    response.onError(Status.INVALID_ARGUMENT.withDescription("You must set at least one of the weather information types.").asException());
                }
                int satisfiedRequestWeatherInformationTypes = 0;
                for (WeatherInformation requestWeatherInformationType : request.getWeatherInformationList()) {
                    if (requestWeatherInformationType == WeatherInformation.Temperature) {
                        if (request.hasMinTemperature() && request.hasMaxTemperature()) {
                            if (request.getMinTemperature() <= responseWeatherInformation.getTemperature() &&
                                request.getMaxTemperature() >= responseWeatherInformation.getTemperature()) {
                                ++satisfiedRequestWeatherInformationTypes;
                            } else {
                                break;
                            }
                        } else if (request.hasMinTemperature()) {
                            if (request.getMinTemperature() <= responseWeatherInformation.getTemperature()) {
                                ++satisfiedRequestWeatherInformationTypes;
                            } else {
                                break;
                            }
                        } else if (request.hasMaxTemperature()) {
                            if (request.getMaxTemperature() >= responseWeatherInformation.getTemperature()) {
                                ++satisfiedRequestWeatherInformationTypes;
                            } else {
                                break;
                            }
                        } else {
                            response.onError(Status.INVALID_ARGUMENT.withDescription("You must set at least one of the following: 'min_temperature', 'max_temperature'.").asException());
                        }
                    } else if (requestWeatherInformationType == WeatherInformation.Humidity) {
                        if (request.hasMinHumidity() && request.hasMaxHumidity()) {
                            if (request.getMinHumidity() <= responseWeatherInformation.getHumidity() &&
                                request.getMaxHumidity() >= responseWeatherInformation.getHumidity()) {
                                ++satisfiedRequestWeatherInformationTypes;
                            } else {
                                break;
                            }
                        } else if (request.hasMinTemperature()) {
                            if (request.getMinHumidity() <= responseWeatherInformation.getHumidity()) {
                                ++satisfiedRequestWeatherInformationTypes;
                            } else {
                                break;
                            }
                        } else if (request.hasMaxHumidity()) {
                            if (request.getMaxHumidity() >= responseWeatherInformation.getHumidity()) {
                                ++satisfiedRequestWeatherInformationTypes;
                            } else {
                                break;
                            }
                        } else {
                            response.onError(Status.INVALID_ARGUMENT.withDescription("You must set at least one of the following: 'min_humidity', 'max_humidity'.").asException());
                        }
                    }
                }
                if (requestWeatherInformationTypes == satisfiedRequestWeatherInformationTypes) {
                    // Simulate server response
                    try {
                        Thread.sleep(ThreadLocalRandom.current().nextInt(5_000, 10_000 + 1));
                        response.onNext(responseWeatherInformation);
                    } catch (InterruptedException exception) {
                        exception.printStackTrace();
                    }
                }

            }
            response.onCompleted();
        }
    }
}
